src/event.rs

```diff
@@ -425,6 +448,118 @@ pub enum WindowEvent {
     RedrawRequested,
 }
 
+/// Represents the pointer type of a pointer event.
+///
+/// ## Platform-specific
+///
+/// **Wayland/X11:** Can only identify [`Touch`](Self::Touch), any other pointer is identified as
+/// [`Mouse`](Self::Mouse).
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum PointerType {
+    Mouse,
+    /// See [`PointerSource::Touch`] for more details.
+    ///
+    /// ## Platform-specific
+    ///
+    /// **macOS:** Unsupported.
+    Touch(FingerId),
+    Unknown,
+}
+
+/// Represents the pointer type and its data for a pointer event.
+///
+/// ## Platform-specific
+///
+/// **Wayland/X11:** Can only identify [`Touch`](Self::Touch), any other pointer is identified as
+/// [`Mouse`](Self::Mouse).
+#[derive(Clone, Debug, PartialEq)]
+pub enum PointerSource {
+    Mouse,
+    /// Represents a touch event
+    ///
+    /// Every time the user touches the screen, a [`WindowEvent::PointerEntered`] and a
+    /// [`WindowEvent::PointerButton`] with [`ElementState::Pressed`] event with an unique
+    /// identifier for the finger is emitted. When a finger is lifted, a
+    /// [`WindowEvent::PointerButton`] with [`ElementState::Released`] and a
+    /// [`WindowEvent::PointerLeft`] event is generated with the same [`FingerId`].
+    ///
+    /// After a [`WindowEvent::PointerEntered`] event has been emitted, there may be zero or more
+    /// [`WindowEvent::PointerMoved`] events when the finger is moved or the touch pressure
+    /// changes.
+    ///
+    /// A [`WindowEvent::PointerLeft`] without a [`WindowEvent::PointerButton`] with
+    /// [`ElementState::Released`] event is emitted when the system has canceled tracking this
+    /// touch, such as when the window loses focus, or on mobile devices if the user moves the
+    /// device against their face.
```

Comment:
This behaviour of getting `Pressed` without `Released` when a touch phase is cancelled, is IMO quite subtle. Perhaps we should rather expose `PointerCancel`?

Reply:
We have attempted to resolve this issue in *length* the other meeting, the outcome was not satisfactory, we should bring it up next meeting as well maybe together we can come up with a better solution.

In the previous proposal we had indeed `TouchCancel`, because it only applies to touch.
It might be a good idea indeed to implement `PointerCancel` and actually use it for other pointer types apart from touch?

This is definitely a weakness of this proposal, if we can't address it in this PR I will open an issue for it.

Reply:
Apologies, summarizing the context:

The reason why we couldn't come up with a good idea/didn't like `TouchCancel` is because it actually goes against the idea of the unified pointer model, where users don't need special handling to correctly handle any pointer type.

So folding it into `PointerLeft` without `PointerButton` with `Released` would make this "natural".
However for users who actually want to handle touch evens explicitly, its harder to make out `TouchCancel`.

So a `PointerCancel` event makes much more sense, because it solves this issue.
But I don't know how we can use it for anything else apart from touch.
However, that said, we can just use it for touch only, as long as users handle it regardless of pointer type, it should fulfill the original purpose.

I like it so far.

Reply:
On another thought, this is also how Web works, another reason to switch to this new event.

Reply:
I'm fine with leaving this to be figure out later

Reply:
I mean, your `PointerCancel` is `PointerLeave` because it means that you effectively lost the focus and thus shouldn't do anything in reaction to pointer input, it's the same for mouse/tablet and touch cancel being here makes sense, at least to me, since I was the one suggesting to do it that way.

Maybe it's not clear what `PointerLeave` actually does? Since what it says that you _lost focus_ it's not like you left the surface since it won't be emitted in case you have on-going grabs, so it's effectively a _cancel_ of the events, just with position attached.

If we add a generic `PointerCancel` it won't be any different than `PointerLeave` because that's what it does.

Reply:
`PointerLeave` has nothing to do with focus.
While it is true that if _pointer focus_ is in play you won't get a `PointerLeave` until you release the pointer focus, this doesn't mean that any other scenario is not relevant.

`PointerLeave` means the pointer has left the surface, with the exception of pointer focus, in which case you will get `PointerLeave` upon loosing pointer focus if the pointer is not inside the surface.
`PointerCancel` means that _input_ should be discarded.

While in the case of touch input, `PointerLeave` and `PointerCancel` are the same thing, as @madsmtm pointed out, the problem is that it is currently hard to figure out when that actually happens because it depends on _not receiving_ a `PointerButton` with `Released`.
While it is possible that users just assume that `PointerLeave` means both and so they can handle it the same, it is not the same as pointed out above. Users who need to distinguish between these two scenarios will have a harder time doing so without a dedicated event.

I think the discussion should rather be around if this differentiation is actually needed, personally I don't know about that. On the other hand W3C deemed it necessary, so maybe digging there could yield something useful.

Reply:
>PointerLeave has nothing to do with focus.

pointer focus, sorry.

>PointerLeave means the pointer has left the surface, with the exception of pointer focus, in which case you will get PointerLeave upon loosing pointer focus if the pointer is not inside the surface.
PointerCancel means that input should be discarded.

This is not the case for grabs, if you want to do it that way, then you'd need to update all the implementations to account for that, which I don't think worth it, since you generally want a _pointer focus_, which what it models right now, at least on X11/Wayland.

>While it is possible that users just assume that PointerLeave means both and so they can handle it the same, it is not the same as pointed out above. Users who need to distinguish between these two scenarios will have a harder time doing so without a dedicated event.

I'll just say that you can get leave on regular mouse devices without releasing the button meaning that you should cancel out everything.

In general, I don't mind having a _cancel_ since I have a case on wayland where you can _remove the pointer_ all together meaning that you likely have to cancel everything/same for keyboard/touch/etc.

Reply:
> This is not the case for grabs, if you want to do it that way, then you'd need to update all the implementations to account for that, which I don't think worth it, since you generally want a _pointer focus_, which what it models right now, at least on X11/Wayland.

The way I described is how it works currently for Web and X11.
I assume this is how it works for every other backend as well already.
If not we should figure out how it works currently and how we want it to work!

What do you mean with "grab"? Don't you mean "pointer focus" with that?

Reply:
I mean that if you press the button and move the pointer out of the window, you won't get the `leave` until you release the button, the coordinates client will be getting will be outside of the window, but the pointer is still _focusing_ the window, because you're holding the button.

So if you treat _leave_ as _left the surface area_ it means that you  should detect when the mouse is not over the window and sent event that it left, but I'd say that all backends prevent _pointer leave_ until you release everything...

Reply:
Cool, we agree then on what `PointerLeave` means, seems this was a misunderstanding then!